fsspec/spec.py

```diff
@@ -445,7 +445,7 @@ def find(self, path, maxdepth=None, withdirs=False, **kwargs):
             if withdirs:
                 files.update(dirs)
             out.update({info["name"]: info for name, info in files.items()})
-        if self.isfile(path) and path not in out:
+        if not out and self.isfile(path):
```

Comment:
Let me try to explain a bit more. The `path` passed in the find could be:
1. An empty directory.
2. A directory with some files/dirs inside.
3. A file.
4.  Does not exist at all.

So, `walk` works on directories usually, but `fsspec.walk` seems to be flexible to also allow files.
https://github.com/intake/filesystem_spec/blob/0c469bfe0f4ef1e0644777791ec36e87519c6eb2/fsspec/spec.py#L406-L408

Given that, if `out` is not empty, it means either that the path was a directory or the `walk` implementation also returned info for the file itself. In both of those cases, we don't need to check for `isfile`, whereas before we were always checking for it.

If `out` is empty, it's either because the `walk` implementation only works based on directories and does not return for the files, or the directory is empty. It's unfortunate that we cannot distinguish between those two, but better than before.

Reply:
There are still a few issues that I see here with the `find` that'd be good to get rid of but could be discussed separately.
1. `find` and `walk` suppressing `FileNotFoundError`.
2. `find` not returning `info` for the file itself.

Reply:
OK, I got it; so the basic point is that we already do this check in `walk`, while having access to the full listing. Sorry for the slightly odd return API. 

For 2., there are specific problematic implementations? Perhaps we merge this and make a new issue to fix them? Note #651 , a general effort to make implementation-agnostic tests and apply them to many/all backends.

Reply:
Also would be happy to see a PR where `find` does not catch FileNotFound, and see if we depend on that behaviour anywhere.

Reply:
> there are specific problematic implementations?

I was relying on the `AbstractFileSystem.find` for the `ls` command implementation, so it is definitely an issue there.  I had to use ugly logic to fix this.

https://github.com/skshetry/webdav4/blob/3307fd568e9ad8fe59bd96894bde6f2575143c44/src/webdav4/cli.py#L579-L586

> Perhaps we merge this and make a new issue to fix them? 

>  would be happy to see a PR where find does not catch FileNotFound

If you are okay with this, let's merge this. I'll create a few more issues that I got in the way.

Especially, since `find` uses `walk`, it's `walk` where we need to fix that.